### PR TITLE
PyPI: Handle non-pythonhosted formula URLs

### PR DESCRIPTION
**Pull request by woodruffw**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This is take two on #15511.

From the original:

This is another quality of life improvement to our handling of Python packages, unlocked by the switch to `pip install --report --dry-run` from a few weeks ago: `pip install` can accept arbitrary distribution URLs, meaning that we no longer require the top-level to be a `pythonhosted` URL in order to auto-bump its dependencies using `update-python-resources`.

This should result in automation for a handful of formulae that previously required manual resource updating. More indirectly, it improves `brew-pip-audit`'s ability to bulk audit Python packages present in Homebrew formulae, resulting in more automated vulnerability fix PRs as well.

### CODE REVIEW

File: Library/Homebrew/utils/pypi.rb [hunk 1]
```diff
@@ -159,30 +228,18 @@ def self.update_python_resources!(formula, version: nil, package_name: nil, extr
...
 
-    unless main_package.valid_pypi_package?
-      return if ignore_non_pypi_packages
-
-      odie "\"#{main_package}\" is not available on PyPI."
+    if version.present?
+      if main_package.valid_pypi_package?
+        main_package.version = version
+      else
+        return if ignore_non_pypi_packages
```
Comment (woodruffw) [inline Library/Homebrew/utils/pypi.rb:238]: This is the only change between this PR and #15511: some callers like `bump-formula-pr` expect to be able to call `update_python_resources!` on arbitrary formulae and have it exit silently.

**Comment by woodruffw** [comment]
CC @p-linnane in particular: I believe I've covered the error mode you noticed, but I'm not super familiar with the various bump workflows these days 🙂

**Review by p-linnane — APPROVED** [review]
Verified that the error I was seeing last time has been resolved. LGTM, thanks @woodruffw!

**Comment by woodruffw** [comment]
CI failures here appear to be flakes.

**Review by MikeMcQuaid — APPROVED** [review]
Looks good, thanks again @woodruffw!

**Comment by woodruffw** [comment]
Thank you both! Fingers crossed on breakage this time 🙂

**Pull request merged**